### PR TITLE
Add first wave of support for building on FreeBSD

### PR DESCRIPTION
**Pull request by zi0r**
This is an initial set of patches to get osquery to build under FreeBSD.  They still need some work:
- the python->python2 change should actually be $(PYTHON) instead.  CMake should check to see if $(PYTHON) is defined, if not, set it to python.
- CMAKE_CXX_COMPILER/CMAKE_C_COMPILER should be able to be overridden.  The FreeBSD ports framework wants to be able to choose the compiler it uses, given a set of requirements.
- Make the calls out to 'git' optional.  In these patches, I've just hardcoded the calls to return "unknown-version"
- If we don't need bash, it would be better to use /bin/sh.  I did not test to see what specifically required bash.  This would be one less dependency.
- The way I'm pulling in the third-party directory is kind of a hack, but it works.  If you guys have a better way to integrate these deps in the future, that would be ideal.  FreeBSD's ports tree has glog, gflags and a version of sqlite3 in it.  Perhaps we can investigate using these existing external dependencies to start work on eliminating the third-party directory wholesale.

My work so far on the port can be found here: https://people.freebsd.org/~zi/osquery.shar

Simply download it and run it to extract the port contents.  Then cd osquery && make
This will pull in all of the dependencies, apply the patches I have for 1.4.5 and get you to the point where linking fails for osqueryd due to libboost_regex/rocksdb issues.
Once we've got a successful build, I'll do some additional review and then commit the port to the FreeBSD ports repo.

Current Problems:
- third-party/glog requires these patches to build: https://svnweb.freebsd.org/ports/head/devel/glog/files/
- Linking fails with a bunch of 'undefined symbol' errors.  This affects libboost_regex and rocksdb.  Switching the linker to use the shared libraries seems to fix this.  This solution might be a good temporary fix, but I suspect we will want to statically link as much as possible, especially given the large dependency list for osquery.
- There might be others that will be discovered after the linking issue is resolved.  I'll continue work as soon as I've got a CMake-fix for it.

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

**Comment by theopolis** [comment]
ok to test

File: CMake/CMakeLibs.cmake [hunk 1]
```diff
@@ -15,7 +15,7 @@ endmacro(SET_OSQUERY_COMPILE)
 
 macro(ADD_OSQUERY_PYTHON_TEST TEST_NAME SOURCE)
   add_test(NAME python_${TEST_NAME}
-    COMMAND python "${CMAKE_SOURCE_DIR}/tools/tests/${SOURCE}" --build "${CMAKE_BINARY_DIR}"
+    COMMAND python2 "${CMAKE_SOURCE_DIR}/tools/tests/${SOURCE}" --build "${CMAKE_BINARY_DIR}"
```
Comment (theopolis) [inline CMake/CMakeLibs.cmake]: I could see this messing up if your pip is pip3, and you install our python requirements into 3, then we call python2 = missing libs. Though the thrift version we're using has string/bytes problems in python3 so we're more or less py2 bound.

File: CMake/FindRocksDB.cmake [hunk 2]
```diff
@@ -73,6 +73,8 @@ if(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
 elseif(${CMAKE_SYSTEM_NAME} MATCHES "Linux")
   # On Linux
   set(ROCKSDB_LIBRARY_NAMES ${ROCKSDB_LIBRARY_NAMES} librocksdb.so)
+elseif(${CMAKE_SYSTEM_NAME} MATCHES "FreeBSD")
+  set(ROCKSDB_LIBRARY_NAMES ${ROCKSDB_LIBRARY_NAMES} librocksdb.so)
```
Comment (theopolis) [inline CMake/FindRocksDB.cmake]: We don't need these variables. These find modules are mostly drop ins from the respective projects. In the second-level CMake we're only pulling in the libs&switches in the static variable: https://github.com/facebook/osquery/blob/master/osquery/CMakeLists.txt#L15

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,7 +1,7 @@
 cmake_minimum_required(VERSION 2.8.12)
 
-set(CMAKE_C_COMPILER "clang")
-set(CMAKE_CXX_COMPILER "clang++")
+#set(CMAKE_C_COMPILER "cc")
+#set(CMAKE_CXX_COMPILER "c++")
```
Comment (theopolis) [inline CMakeLists.txt]: But I REALLY want LLVM ;) 

Let me think of the best way to allow a compiler override.

File: CMakeLists.txt [hunk 4]
```diff
@@ -139,7 +140,7 @@ execute_process(
 if(DEFINED ENV{SDK_VERSION})
   set(OSQUERY_BUILD_SDK_VERSION "${ENV{SDK_VERSION}}")
 else()
-  string(REPLACE "-" ";" OSQUERY_BUILD_SDK_VERSION ${OSQUERY_BUILD_VERSION})
+  string(REPLACE "-" ";" OSQUERY_BUILD_SDK_VERSION "1.4.5")
```
Comment (theopolis) [inline CMakeLists.txt]: Is there any way for ports to set this variable? And then send it in to CMake within CMAKE_ARGS? Let me know what you think is an appropriate variable name to check. OSQUERY_EXTERNAL_VERSION, OSQUERY_MANUAL_VERSION? Something along those lines.

CMake can check for environment variables too. Either way, if it's set then CMake will skip trying to discover through git tags.

Comment (zi0r) [inline CMakeLists.txt]: Yes, I can easily add -DFOO=bar to CMAKE_ARGS for anything.

The port is aware of its version, so I can literally do something like:
CMAKE_ARGS+=-DOSQUERY_EXTERNAL_VERSION="${PORTVERSION}"

File: CMake/CMakeLibs.cmake (hunk 1 above)
Comment (zi0r) [inline CMake/CMakeLibs.cmake]: This was just a quick hack to make it go.  It should really become $(PYTHON) so we can override it from the port.  Then, on the osquery side, if it isn't defined, default to "python"

**Pull request closed without merging**